### PR TITLE
#36020 Improved support for immutable plugin configs

### PR DESCRIPTION
**Pull request by manneohrstrom**
When a plugin config is defined as `baked` in the plugin manifest, we know that it is not going to change or be updated (like the case with for example RV). This change improves this workflow and avoids doing a setup of such a configuration in temp space prior to runtime. Instead, the config is baked out at build time and runs directly from within the built bundle cache itself. Details:

- The `bootstrap.Configuration` class has been split into two - `CachedConfiguration` for the traditional workflow where a config is created on the fly in temp space (our standard use case) and a `BakedConfiguration` which is used you have a baked, immutable config.
- The various I/O operations around building configs has been split out into a separate `ConfigurationWriter` class for convenience.
- The `baked` descriptor type already introduced in the build script is extended to carry through into the built data - for these baked descriptors, a complete and installed configuration is baked in:

image

- When a baked config is detected, runtime parameters such as project id, pipeline config id, that are normally baked into the config in files are passed via a pickled environment variable.
- Fixed a bug in the `filesystem.copy_file` method: now properly handles the case when the skip list is cleared.
- `sgtk.util.shotgun.get_associated_sg_base_url()` uses the authenticated user's base url if available and only falls back on `shotgun.yml` if needed.
- Build script outputs file sizes.

### CODE REVIEW

File: python/tank/bootstrap/baked_configuration.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                  LOCAL_CFG_DIFFERENT, or LOCAL_CFG_INVALID
+        """
+        return self.LOCAL_CFG_UP_TO_DATE
+
+    def update_configuration(self):
+        """
+        Ensure that the configuration is up to date with the one
+        given by the associated descriptor.
+
+        This method fails gracefully and attempts to roll back to a
```
Comment (josh-t) [inline python/tank/bootstrap/baked_configuration.py]: This sentence should probably be removed. Perhaps update this docstring entirely noting that for baked configs, this is a no-op.

File: python/tank/bootstrap/cached_configuration.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+                log.debug("Previous config restore complete...")
+
+                log.debug("Restoring previous core %s" % core_backup_path)
+                filesystem.copy_folder(
+                    core_backup_path,
+                    os.path.join(self._path.current_os, "install", "core")
+                )
+                log.debug("Previous core restore complete...")
+
+        # @todo - prime caches (yaml, path cache)
```
Comment (josh-t) [inline python/tank/bootstrap/cached_configuration.py]: thought i'd point this out just in case it needs to happen before release.

File: python/tank/bootstrap/configuration_writer.py [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+            "core",
+            constants.PIPELINECONFIG_FILE
+        )
+
+        with self._open_auto_created_yml(pipeline_config_path) as fh:
+            yaml.safe_dump(pipeline_config_content, fh)
+            fh.write("\n")
+            fh.write("# End of file.\n")
+
+    def write_bare_pipeline_config_file(self, entry_point):
```
Comment (josh-t) [inline python/tank/bootstrap/configuration_writer.py]: The contents of this method are pretty similar to the end of the previous method. Maybe there's one method to write a config and these two methods just pass in the content dict?

File: python/tank/pipelineconfig.py [hunk 4]
```diff
@@ -109,8 +108,28 @@ def __init__(self, pipeline_configuration_path):
         else:
             self._bundle_cache_fallback_paths = []
 
-        # Populate the global yaml_cache if we find a pickled cache
-        # on disk.
+        # look for environment variables
+        if constants.ENV_VAR_EXTERNAL_PIPELINE_CONFIG_DATA in os.environ:
```
Comment (josh-t) [inline python/tank/pipelineconfig.py]: Values here are overriding values in the metadata file. Maybe a quick note about why that is?

File: scripts/build_plugin.py [hunk 5]
```diff
@@ -64,6 +66,37 @@ class OptionParserLineBreakingEpilog(optparse.OptionParser):
     def format_epilog(self, formatter):
         return self.epilog
 
+def _compute_size_kbytes(path):
```
Comment (josh-t) [inline scripts/build_plugin.py]: perhaps this could/should live in `tank.util.filesystem`?

File: scripts/build_plugin.py [hunk 6]
```diff
@@ -64,6 +66,37 @@ class OptionParserLineBreakingEpilog(optparse.OptionParser):
...
+    :param desc_type: Descriptor.ENGINE | Descriptor.APP | Descriptor.FRAMEWORK
+    :param desc_dict: descriptor dict or uri
+    :param target_path: bundle cache root to cache into
+    """
+    desc = create_descriptor(sg, desc_type, desc_dict)
+    desc.ensure_local()
+    desc_size_kb = _compute_size_kbytes(desc.get_path())
+    logger.info("Caching %s into plugin bundle cache (size %d KiB)" % (desc, desc_size_kb))
+    if not desc._io_descriptor.is_immutable():
+        logger.warning("Descriptor %r may not work for other users using the plugin!" % desc)
```
Comment (josh-t) [inline scripts/build_plugin.py]: This part is great. Really helpful warning when developing.

File: scripts/build_plugin.py [hunk 7]
```diff
@@ -322,24 +353,40 @@ def _bake_manifest(manifest_data, config_uri, core_descriptor, plugin_root):
...
+                fh.write("\n\n")
+                fh.write("def get_sgtk_pythonpath(plugin_root):\n")
+                fh.write("    import os\n")
+                fh.write("    return os.path.join(plugin_root, %s)\n" %
+                         ", ".join('"%s"' % dir for dir in core_path_relative_parts))
+                fh.write("\n\n")
+
+            else:
+                # the core descriptor is outside of bundle cache
+                logger.warning("Your core %r has its payload outside the plugin bundle cache. "
```
Comment (josh-t) [inline scripts/build_plugin.py]: another great warning.

File: scripts/build_plugin.py [hunk 8]
```diff
@@ -322,24 +353,40 @@ def _bake_manifest(manifest_data, config_uri, core_descriptor, plugin_root):
...
+                               "This plugin cannot be distributed to others." % core_descriptor)
+
+                core_path_parts = os.path.normpath(core_descriptor.get_path()).split(os.path.sep)
+                core_path_parts.append("python")
+
+                fh.write("\n\n")
+                fh.write("def get_sgtk_pythonpath(plugin_root):\n")
+                fh.write("    return '%s'\n" % os.path.sep.join(core_path_parts))
```
Comment (josh-t) [inline scripts/build_plugin.py]: maybe add a comment in here about why the `plugin_root` isn't used just to reenforce for developers? I guess echoing the warning above?

File: python/tank/bootstrap/baked_configuration.py (hunk 1 above)
Comment (manneohrstrom) [inline python/tank/bootstrap/baked_configuration.py]: ah yes! fixed!

File: python/tank/bootstrap/cached_configuration.py (hunk 2 above)
Comment (manneohrstrom) [inline python/tank/bootstrap/cached_configuration.py]: that's fine - it's purely an opportunity to improve performance now that we have the local bootstrap. We got a separate ticket for it, but it's lower prio!

File: python/tank/bootstrap/configuration_writer.py (hunk 3 above)
Comment (manneohrstrom) [inline python/tank/bootstrap/configuration_writer.py]: good point! fixed!

File: python/tank/pipelineconfig.py (hunk 4 above)
Comment (manneohrstrom) [inline python/tank/pipelineconfig.py]: yes! fixed!

File: scripts/build_plugin.py (hunk 5 above)
Comment (manneohrstrom) [inline scripts/build_plugin.py]: done!

File: scripts/build_plugin.py (hunk 8 above)
Comment (manneohrstrom) [inline scripts/build_plugin.py]: great. done!

File: developer/build_plugin.py [hunk 9]
```diff
@@ -454,23 +537,94 @@ def main():
     Handles argument parsing and validation and then calls the script payload.
     """
 
-    usage = "%prog source_path target_path (run with --help for more information)"
+    usage = "%prog source_path [-dskc] target_path (run with --help for more information)"
```
Comment (josh-t) [inline developer/build_plugin.py]: The `[-dskc]` part here is a little misleading. This makes it look like flags. Maybe replace with something like `[options]` since otherwise it could get long? Or maybe just insert full option and make it long?

File: developer/build_plugin.py [hunk 10]
```diff
@@ -454,23 +537,94 @@ def main():
...
+        "against any shotgun site. By default, it will use the toolkit authentication APIs stored "
+        "credentials, and if such are not found, it will prompt for site, username and password."
+    )
+
+    group.add_option(
+        "-s",
+        "--shotgun-host",
+        default=None,
+        action="store",
+        help="Shotgun Script to use to authenticate to retrieve app store credentials."
```
Comment (josh-t) [inline developer/build_plugin.py]: Looks like the wrong string here.

File: developer/build_plugin.py [hunk 11]
```diff
@@ -454,23 +537,94 @@ def main():
...
+    )
+
+    group.add_option(
+        "-n",
+        "--shotgun-script-name",
+        default=None,
+        action="store",
+        help="Shotgun Script to use to authenticate to retrieve app store credentials."
```
Comment (josh-t) [inline developer/build_plugin.py]: Maybe reword this `help` string. Suggestion:

`help="Script name to use for Shotgun authentication"`

File: developer/build_plugin.py [hunk 12]
```diff
@@ -454,23 +537,94 @@ def main():
...
+    )
+
+    group.add_option(
+        "-k",
+        "--shotgun-script-key",
+        default=None,
+        action="store",
+        help="Shotgun Script key to use to authenticate to retrieve app store crendentials."
```
Comment (josh-t) [inline developer/build_plugin.py]: Maybe reword this `help` string. Suggestion:

`help="Script key to use for Shotgun authentication"`

File: developer/build_plugin.py [hunk 13]
```diff
@@ -495,11 +655,29 @@ def main():
...
+        sg_user = sg_auth.create_script_user(script_name, script_key, options.shotgun_host)
+
+    else:
+        # get user, prompt if necessary
+        sg_user = sg_auth.get_user()
+
     sg_connection = sg_user.create_sg_connection()
+    # make sure we are properly connected
+
+    sg_connection.find_one("HumanUser", [])
```
Comment (josh-t) [inline developer/build_plugin.py]: If this is testing the connection, maybe it should be wrapped in a `try/except`?

File: developer/build_plugin.py [hunk 14]
```diff
@@ -312,34 +331,92 @@ def _bake_manifest(manifest_data, config_uri, core_descriptor, plugin_root):
...
-                     ", ".join('"%s"' % dir for dir in core_path_relative_parts))
-            fh.write("\n\n")
 
+            # set base configuration
+            fh.write("    manager.base_configuration = '%s'\n" % config_uri)
+
+            # set entry point
+            fh.write("    manager.entry_point = '%s'\n" % manifest_data["entry_point"])
+
+            # set shotgun config look flag if defined
```
Comment (josh-t) [inline developer/build_plugin.py]: minor: `lookup`

**Comment by manneohrstrom** [comment]
merged in master.

File: developer/build_plugin.py (hunk 9 above)
Comment (manneohrstrom) [inline developer/build_plugin.py]: changed to [options]!

File: developer/build_plugin.py (hunk 10 above)
Comment (manneohrstrom) [inline developer/build_plugin.py]: ekk! sorry, fixed! shouldn't code in the night... ;(

Comment (manneohrstrom) [inline developer/build_plugin.py]: Help output is now:

```
Usage: build_plugin.py [options] source_path target_path

Builds a standard toolkit plugin structure ready for testing and deploy

Options:
  -h, --help            show this help message and exit
  -d, --debug           Enable debug logging
  -c BOOTSTRAP_CORE_URI, --bootstrap-core-uri=BOOTSTRAP_CORE_URI
                        Specify which version of core to be used by the
                        bootstrap process. If not specified, defaults to the
                        most recently released core.

  Shotgun Authentication:
    In order to download content from the Toolkit app store, the script
    will need to authenticate against any shotgun site. By default, it
    will use the toolkit authentication APIs stored credentials, and if
    such are not found, it will prompt for site, username and password.

    -s SHOTGUN_HOST, --shotgun-host=SHOTGUN_HOST
                        Shotgun host to authenticate with.
    -n SHOTGUN_SCRIPT_NAME, --shotgun-script-name=SHOTGUN_SCRIPT_NAME
                        Script to use to authenticate with the given host.
    -k SHOTGUN_SCRIPT_KEY, --shotgun-script-key=SHOTGUN_SCRIPT_KEY
                        Script key to use to authenticate with the given host.


Details and Examples
--------------------

In its simplest form, just provide a source and target folder for the build.

> python build_plugin.py ~/dev/tk-maya/plugins/basic /tmp/maya-plugin

For automated build setups, you can provide a specific shotgun API script name and
and corresponding script key:

> python build_plugin.py
            --shotgun-host='https://mysite.shotgunstudio.com'
            --shotgun-script-name='plugin_build'
            --shotgun-script-key='<script-key-here>'
            ~/dev/tk-maya/plugins/basic /tmp/maya-plugin

By default, the build script will use the latest app store core for its bootstrapping.
If you want to use a specific core for the bootstrap, this can be specified via the
--bootstrap-core-uri option:

> python build_plugin.py
            --bootstrap-core-uri='sgtk:descriptor:dev?path=~/dev/tk-core'
            ~/dev/tk-maya/plugins/basic /tmp/maya-plugin

For information about the various descriptors that can be used, see
http://developer.shotgunsoftware.com/tk-core/descriptor

```

File: developer/build_plugin.py (hunk 13 above)
Comment (manneohrstrom) [inline developer/build_plugin.py]: done!
